cpp/run3ScoutingLooper.C

```diff
@@ -0,0 +1,587 @@
+      PV_x = 0; PV_y = 0; PV_z = 0;
+
+      if (pvs.size() > 0 && pvs[0].isValidVtx()) {
+        PV_x = pvs[0].x();
+        PV_y = pvs[0].y();
+        PV_z = pvs[0].z();
+      }
+
+      // SV selection
+      auto svs = getObject<std::vector<Run3ScoutingVertex>>(ev, "hltScoutingMuonPacker", "displacedVtx");
+      SVs.clear();
+
+      unsigned int nSVs = svs.size();
+      if (nSVs < 1)
+        continue;
+
+      for (unsigned int iSV=0; iSV<nSVs; ++iSV) {
+        auto sv = svs[iSV];
+        if (!(sv.isValidVtx()))
+          continue;
+
+        SVs.index.push_back(iSV);
+        SVs.ndof.push_back(sv.ndof());
+        SVs.x.push_back(sv.x());
+        SVs.y.push_back(sv.y());
+        SVs.z.push_back(sv.z());
+        SVs.xe.push_back(sv.xError());
+        SVs.ye.push_back(sv.yError());
+        SVs.ze.push_back(sv.zError());
+        SVs.chi2.push_back(sv.chi2());
+        SVs.prob.push_back(TMath::Prob(sv.chi2(), sv.ndof()));
+        SVs.chi2Ndof.push_back(sv.chi2()/sv.ndof());
+        float lxy = TMath::Sqrt((sv.x()-PV_x)*(sv.x()-PV_x)+(sv.y()-PV_y)*(sv.y()-PV_y));
+        SVs.lxy.push_back(lxy);
+        SVs.l3d.push_back(TMath::Sqrt(lxy*lxy+(sv.z()-PV_z)*(sv.z()-PV_z)));
+
+        float mindx=1e6, mindy=1e6, mindz=1e6, mindxy=1e6, mind3d=1e6;
+        float maxdx=-1, maxdy=-1, maxdz=-1, maxdxy=-1, maxd3d=-1;
+        for (unsigned int jSV=0; jSV<nSVs; ++jSV) {
+          if (jSV==iSV)
+            continue;
+          auto svOther = svs[jSV];
+          if (!(svOther.isValidVtx()))
+            continue;
+
+          float dx = (sv.x()-svOther.x())*(sv.x()-svOther.x()); // Squared for the moment
+          float dy = (sv.y()-svOther.y())*(sv.y()-svOther.y()); // Squared for the moment
+          float dz = (sv.z()-svOther.z())*(sv.z()-svOther.z()); // Squared for the moment
+          float dxy = TMath::Sqrt(dx+dy);
+          float d3d = TMath::Sqrt(dx+dy+dz);
+          dx = TMath::Sqrt(dx); // Back to the proper value
+          dy = TMath::Sqrt(dy); // Back to the proper value
+          dz = TMath::Sqrt(dz); // Back to the proper value
+
+          if (dx<mindx) mindx = dx;
+          if (dx>maxdx) maxdx = dx;
+          if (dy<mindy) mindy = dy;
+          if (dy>maxdy) maxdy = dy;
+          if (dz<mindz) mindz = dz;
+          if (dz>maxdz) maxdz = dz;
+          if (dxy<mindxy) mindxy = dxy;
+          if (dxy>maxdxy) maxdxy = dxy;
+          if (d3d<mind3d) mind3d = d3d;
+          if (d3d>maxd3d) maxd3d = d3d;
+        }
+        SVs.mindx.push_back(mindx);
+        SVs.mindy.push_back(mindy);
+        SVs.mindz.push_back(mindz);
+        SVs.mindxy.push_back(mindxy);
+        SVs.mind3d.push_back(mind3d);
+        SVs.maxdx.push_back(maxdx);
+        SVs.maxdy.push_back(maxdy);
+        SVs.maxdz.push_back(maxdz);
+        SVs.maxdxy.push_back(maxdxy);
+        SVs.maxd3d.push_back(maxd3d);
+      }
+      SVs.sort();
```

Comment:
After sorting the SVs, should we already select "good" ones and identify overlapping SVs?
For selection, it should be fair to use the same selection used in EXO-20-014: "xe<0.05 && ye<0.05 && ze<0.10 && chi2/ndof<5.0".
For identifying overlaps among 'good' ones, could use "dx(SVi,SVj)<0.05 && dy(SVi,SVj)<0.05 && dz(SVi,SVj)<0.1 && dxy(SVi,SVj)<0.05 && d3d<0.1", which reflects the selection applied in EXO-20-014.
If overlapping SVs are found, we can then store "nSVoverlap", vector<int> "SVoverlap_vtxIdxs" (to store the actual indices of overlapping SVs), "SVoverlap_x", "SVoverlap_y", ""SVoverlap_z", "SVoverlap_lxy", "SVoverlap_l3d", where "x", "y", and "z" can be obtained as the weighted average of the corresponding coordinates across "nSVoverlap", with weight = chi2 probability (the highest gets more weight); then, "lxy" and "l3d" can be recomputed accordingly.
Then, for each muon we can already assign the best SV:
- if nSVoverlap>0 and the index of the highest-ranked SV associated to the muon is in SVoverlap_vtxIdxs,  we consider the overlap SV as the best SV;
- otherwise,  we consider the highest-ranked SV as the best SV.
Given the best SV/muon, we can then already correct the muon transverse IP and the muon φ:
- for IP: https://github.com/aminnj/scouting/blob/master/batch/babymaker.py#L1179
- for φ: https://github.com/aminnj/scouting/blob/master/batch/babymaker.py#L584, which makes use of https://github.com/aminnj/scouting/blob/master/batch/data/propagation_utils.cc

In principle, we could then also recompute the expected number of (pixel and/or tracker) hits, as in:
https://github.com/aminnj/scouting/blob/master/batch/Scouting/NtupleMaker/plugins/HitMaker.cc
However, this is less trivial.

Similarly, we could assess whether the SV is a material vertex, or not, like in:
https://github.com/aminnj/scouting/blob/master/batch/babymaker.py#L945
(which is defined in https://github.com/aminnj/scouting/blob/master/batch/babymaker.py#L571).
However, this is also not trivial.

Furthermore, the last two points are less trivial than for EXO-20-014, since the 'problem' now extends beyond the pixel tracker.

Reply:
> After sorting the SVs, should we already select "good" ones and identify overlapping SVs?
For selection, it should be fair to use the same selection used in EXO-20-014: "xe<0.05 && ye<0.05 && ze<0.10 && chi2/ndof<5.0".

I agree. But I don't see a reason why to do this after we sort. In fact, I would just do it in the SV loop. That would mean that we won't store SVs that don't pass this selection - do we care about those?

For all the rest, I will start implementing them already.